### PR TITLE
Improve gyro LPF handling for small copters

### PR DESCRIPTION
**Pull request by andyp1per**
AP_InertialSensor: Replace gyro LPF with two independently configurable PT1s.

BF 4.0 testing shows that best results for filtering the absence of dynamic
filtering is two single-pass filters with different cutoff frequencies.
The new filter is applied first and configured via INS_GYRO_FILT2. If the value
is 0 then it tracks the value of INS_GYRO_FILT.

### CODE REVIEW

File: libraries/AP_InertialSensor/AP_InertialSensor.cpp [hunk 1]
```diff
@@ -453,6 +453,14 @@ const AP_Param::GroupInfo AP_InertialSensor::var_info[] = {
     // @Bitmask: 0:FirstIMU,1:SecondIMU,2:ThirdIMU
     AP_GROUPINFO("ENABLE_MASK",  40, AP_InertialSensor, _enable_mask, 0x7F),
 
+    // @Param: GYRO_FILTER2
+    // @DisplayName: Second Gyro filter cutoff frequency
+    // @Description: Filter cutoff frequency for gyroscopes. This can be set to a lower value to try to cope with very high vibration levels in aircraft. This option takes effect on the next reboot. A value of zero means track the value of GYRO_FILTER
```
Comment (amilcarlucas) [inline libraries/AP_InertialSensor/AP_InertialSensor.cpp]: please add the needs_reboot atribute.

Comment (andyp1per) [reply libraries/AP_InertialSensor/AP_InertialSensor.cpp]: I don't think it needs this does it? The code has been written to detect changes on the fly and update the filter. The same is true for INS_GYRO_FILT

Comment (amilcarlucas) [reply libraries/AP_InertialSensor/AP_InertialSensor.cpp]: If that is the case then update the description text

Comment (andyp1per) [reply libraries/AP_InertialSensor/AP_InertialSensor.cpp]: Ah good catch

File: libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp [hunk 2]
```diff
@@ -204,8 +204,10 @@ void AP_InertialSensor_Backend::_notify_new_gyro_raw_sample(uint8_t instance,
         _imu._last_delta_angle[instance] = delta_angle;
         _imu._last_raw_gyro[instance] = gyro;
 
-        // apply the low pass filter
-        _imu._gyro_filtered[instance] = _imu._gyro_filter[instance].apply(gyro);
+        // apply the low pass filters
+        _imu._gyro_filtered[instance] = _imu._gyro_filter2[instance].apply(gyro);
+        _imu._gyro_filtered[instance] = _imu._gyro_filter[instance].apply(_imu._gyro_filtered[instance]);
```
Comment (magicrub) [inline libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp]: have you done any tests for how this effects the phase and time delay? This second order filter may have some interesting effects when it hits the EKF.

Comment (andyp1per) [reply libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp]: I think the EKF uses unfiltered delta angles, so I believe unaffected by this change.

Comment (andyp1per) [reply libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp]: Also to be clear this replaces a 2-pass filter, with two 1-pass filters so I think time delay should be the same.

**Comment by sh83** [comment]
Actually there is another 188hz lpf and 8kz to 1khz rate reduction at AP_InertialSensor_Invensense.cpp. So after this change we will have 3 lp filters.

**Comment by andyp1per** [comment]
In BF the HW LPF is configurable and generally they switch it off - I guess it's not a very useful frequency especially as at high throttle they want the LPF's up high.
I assume PT1 is applied on gyro side - but it doesn't much matter since they commonly run the PID loop at 8k as well. Or sometimes 8k gyro/4k PID. Interestingly according to ctsnooze the loop time is not as significant on latency as you might think - it much more about filter target frequency. So for instance a wide notch has a big effect of latency. Similarly having a strong LPF at low frequencies has a big(er) effect. A typical config in BF is LPF's set at 100Hz and 300Hz. Note that one thing this change allows you to do is remove a filter by setting INS_GYRO_FILT to 0 and only using INS_GYRO_FILT2 - you then have a PT1 instead of a BIQUAD. I have flown with this fine.

**Comment by andyp1per** [comment]
Also interestingly they find they don't benefit from 32k sampling much as it introduces too much noise and therefore requires stronger filterings, introducing more latency.

**Comment by andyp1per** [comment]
One thing that occurs to me is that I think filters should be applied from high freq to low so that the higher freq filters are seeing a cleaner signal to work with. In a typical config that probably means that it should go FILT2 -> NOTCH -> FILT with NOTCH sitting at the motor peak (around 200Hz on a typical small quad)

**Comment by andyp1per** [comment]
I think there is actually a bug in the HW LPF we set. It's highly confusing because the definitions for the gyro config for the invensense chips has changed. For 20649 and family we are doing full rate sampling (9k) with essentially no filtering (12k) with fast sampling. For 9250 and 20608 we are doing 8k sampling with 256Hz filter (LPF config option 0) with fast sampling - this is **not** LPF disabled. For effectively disabling the LPF you have to use option 7 which gives 8k sampling 3k LPF. The BF guys have this as an "experimental" option and it may not matter as I suspect they do 32k sampling which does mean option 0 disables LPF.

**Comment by andyp1per** [comment]
So I tried changing this and it works pretty well. If anything I would say the copter feels more solid in the air. I also note that we are inconsistent with the accels as well where we do full rate sampling with no filtering. I think the right thing to do is switch this off and do it all in software, but it's possible that may require some changes to the EKF which thinks it's using "unfiltered" values, when in essence it is not.

**Comment by lthall** [comment]
Great work going over all the core settings here Andy!!

Ok, so I have been going over this concept and this reduces the performance of the filter considerably.

When using this approach on our standard setups, this results in significant attenuation below the cutoff frequency compared to the 2p filter. It also increases the latency  by about 40% at low frequencies. So this would result in all well tuned aircraft to be retuned and they would never get the same level of performance.

With the same low frequency latency a 2p filter will result in about 20% less attenuation in the frequency range below the filter settings (at the higher end). It will result in the same attenuation of noise above the filter cut off at approximately 140% the cut off frequency. After that it quickly moves to a 50% improvement in noise rejection. 

So in some extremely specific examples you may get a 20% improvement in attenuation but in all other cases it results in more latency and less attenuation of noise.

By setting the 2p filter frequency correctly you should be able to better performance in almost all cases when compared to two 1p filters.

I think the main reason this is useful for BF is ease of use. "If you have low noise levels you can turn off the first filter".

**Comment by andyp1per** [comment]
Thanks for the analysis @lthall ! So if less filtering was beneficial simply being able to select the filter might work better.

**Comment by lthall** [comment]
Hi Andy, yes that would be correct. A better approach would be to let the user choose a 2p or a 1p filter. The default would not change then. To justify this change we would need to show that a 1p filter does provide an advantage and clearly define when a 1p filter is appropriate. 

While the implementation cost is small there will be an ongoing support cost associated with people asking which setting to use for their aircraft. This can be reduced if we can make it easy to tell when the 1p filter should be used.

However, based on my analysis the 1p filter will never be a better choice than the 2p filter. I am very open to be wrong though!!

**Comment by andyp1per** [comment]
Without further filtering on D I suspect the latency improvement won't be worth it - and as you have pointed out privately the AC control model is different to that of BF, so filtering hard on D will have different effects.

**Comment by andyp1per** [comment]
I'm going to close this in favour of https://github.com/ArduPilot/ardupilot/pull/11795
Testing may reveal some of value here, but it will surely be different

**Pull request closed without merging**